core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,379 @@
+                //  have the same name and it mixes performance stats. See https://github.com/janino-compiler/janino/issues/137
+                long counter = longVal.incrementAndGet();
+                String classTemplate = createClassTemplate(counter, priorityVariables, speedVariables, lookup, customModel);
+                cu = (Java.CompilationUnit) new Parser(new org.codehaus.janino.Scanner("source", new StringReader(classTemplate))).
+                        parseAbstractCompilationUnit();
+                cu = injectStatements(priorityStatements, speedStatements, cu);
+                SimpleCompiler sc = createCompiler(counter, cu);
+                clazz = sc.getClassLoader().loadClass("com.graphhopper.Test" + counter);
+                if (CACHE.size() < CACHE_SIZE)
+                    CACHE.put(key, clazz);
+                else if (DYN_CACHE_SIZE > 0)
+                    DYN_CACHE.put(key, clazz);
+            }
+
+            // The class does not need to be thread-safe as we create an instance per request
+            CustomWeightingHelper prio = (CustomWeightingHelper) clazz.getDeclaredConstructor().newInstance();
+            prio.init(lookup, avgSpeedEnc, customModel.getAreas());
+            return prio;
+        } catch (Exception ex) {
+            String location = "";
+            if (ex instanceof CompileException)
+                location = " in " + ((CompileException) ex).getLocation().getFileName();
+            throw new IllegalArgumentException("Cannot compile expression " + location + ", " + ex.getMessage(), ex);
+        }
+    }
+
+    /**
+     * Parse the expressions from CustomModel relevant for the method getSpeed - see createClassTemplate.
+     *
+     * @return the created statements (parsed expressions)
+     */
+    private static List<Java.BlockStatement> createGetSpeedStatements(Set<String> speedVariables,
+                                                                      CustomModel customModel, EncodedValueLookup lookup,
+                                                                      double globalMaxSpeed, double maxSpeedFallback) throws Exception {
+        List<Java.BlockStatement> speedStatements = new ArrayList<>();
+        speedStatements.addAll(verifyExpressions(new StringBuilder(), "speed_factor_user_statements", speedVariables,
+                customModel.getSpeedFactor(), lookup,
+                num -> "speed *= " + num + ";\n", ""));
+        StringBuilder codeSB = new StringBuilder("boolean applied = false;\n");
+        speedStatements.addAll(verifyExpressions(codeSB, "max_speed_user_statements",
+                speedVariables, customModel.getMaxSpeed(), lookup,
+                num -> "applied = true; speed = Math.min(speed," + num + ");",
+                "if (!applied && speed > " + maxSpeedFallback + ") return " + maxSpeedFallback + ";\n" +
+                        "return Math.min(speed, " + globalMaxSpeed + ");\n"));
```

Comment:
This is kind of hard to read. Can you explain what the `applied` flag does and why there are two speed values (`maxSpeedFallback` and `globalMaxSpeed`)?

Reply:
This is basically what we do in master: https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/weighting/custom/SpeedCalculator.java#L155

Reply:
But I'm unsure if we really need this. I think we could replace maxSpeedFallback with `max_speed` and `first_match`.

Reply:
Ah ok wow this logic is already not so easy to understand in master :) So we have the `max_speed` section we can use to set max speeds and then a global max speed fallback that always applies. Yes it seems like we can set a global max speed fallback using the else clause:

```yaml
max_speed:
   - if: road_class == PRIMARY
     value: 80
   - else: 60
```

Note that here we should call the value `value` (or at least **not** `factor`).
Setting only a fallback (but no other max speeds) works like this:

```yaml
max_speed:
   - if: false
     factor: 999 # does not matter?!
   - else: 60
```

Reply:
> Note that here we should call the value value (or at least not factor).

:+1: 

> Setting only a fallback (but no other max speeds) works like this:

Wouldn't this better work?

```yaml
max_speed:
   - if: true
     value: 60
```

(and was it intentional to use factor in this example?)

Reply:
> Wouldn't this better work?

Oh of course yes :)

> (and was it intentional to use factor in this example?)

No this should also be `value` (or similar)

Reply:
Ok, will remove getMaxSpeedFallback after you merged #2211.

Reply:
Ah you mean in the current format we can already do:

```yaml
max_speed:
   true: 60
```

?

Reply:
Yes

Reply:
Ah, the reason we have this separate entry is that this improves speed measurable. 
(A better estimate in getMinWeight results in less visited nodes)

But instead of the "applied" logic we could just add it (and throw an exception if `true` already exists)?

Reply:
But can we not derive the global maximum speed from the custom model? It's just the maximum over all `max_speed` entries and the encoder's base speed.

Reply:
I think what we need is the smallest (unconditional) max_speed - as we already know the maximum, which is the max_speed of the encoder. And e.g. a value with a condition like `max_speed: { road_class == PRIMARY: 60 }` is not sufficient to lower the maximum speed of the encoder, which is required to improve our getMinWeight estimate.

So we would need to loop over all entries and e.g. compare to `true` (?) and pick the smallest value or allow only one unconditional value? We could document this special thing as a performance improvement and avoid the special handling variable maxSpeedFallback. From the user perspective this would not be more complicated and for us it would improve the situation a bit.

Reply:
> I think what we need is the smallest (unconditional) max_speed

Yes, I was wrong here. We need to find the maximum speed that can occur across *all* edges. Maximum values that only apply to *some* edges do not matter.

> So we would need to loop over all entries and e.g. compare to true (?)

Yes. The maximum speed is the base speed for the encoder **unless** there is some max speed rule that applies to *all* edges, for example one with an else clause or condition `true`.

> and pick the smallest value or allow only one unconditional value?

Right. I think we should only allow one such rule. Otherwise it will depend on the order of the rules which might be not intuitive:

```yaml
max_speed:
    - if: road_class == MOTORWAY
      value: 100
      else: 50
    - if: surface == GRAVEL
      value: 70
      # ouch, this will never be applied anyway, because the road_class rule comes first
      else:  30 
    - if: true
      # same here...
      value: 80
```

> We could document this special thing as a performance improvement 

Can we postpone this performance improvement and use the encoder speed as max speed for now so we can finalize the format and answer the other questions first? Its only an improvement for those custom models that actually reduce the global speed, right?

Reply:
>  for example one with an else clause or condition `true`

Not all else-clauses - only for the else clause of `false` (?)

> Otherwise it will depend on the order of the rules which might be not intuitive:
> same here...

You probably meant an if-clause block like:

```yml
max_speed:
 - if: road_class == MOTORWAY
   value: 100
 - else if: surface == GRAVEL
   value: 70
 - else if: true
   value: 80
```

Because for max_speed the order does not matter as we go through all if expressions and instead of multiplication `speed = factor* speed` we do `speed = Math.min(max_speed, speed)`. In your example the value 80 will still be picked in case of road_class==MOTORWAY.

So if we scan for `if: true` this should work? (And maybe also `if: false -> else`)

> And note that this is different to speed/priority where multiple else clauses can make sense because the corresponding factors all contribute to the combined result.

Why? Can you explain this? Maybe this is the reason I read your above example incorrectly.

In master the max_speed expressions are applied several times.

> Can we postpone this performance improvement

It exists in master (via the max_speed_fallback) and it shouldn't be hard due to your observation above?

> Its only an improvement for those custom models that actually reduce the global speed, right?

I'm unsure the practical examples are improved but a too loose getMinWeight approximation can have a big impact on A* speed. And as it is not uncommon to reduce the overall max_speed e.g. base your work on car and say you can drive 100km/h maximum.

Reply:
Hm yeah this is a bit confusing. Right now this comes down to the following question for me, and before this is clear I think I cannot answer your questions properly. Let's say we have a custom model with these 'rules.

```yaml
- road_class == MOTORWAY --> 100
- surface == GRAVEL --> 70
- road_environment == TUNNEL --> 80
```

For an edge that has all three attributes what do we expect:

1) 100, because this is the first rule that matches (order matters)
2) 80, because it is the last rule that matches (order matters). this sounds weird, but that would be the result if we had single if statements with no return (see below)
3) 70, because out of all matches this is the most restrictive speed limit (I think this is done in master currently)
4) 100, because we take the most restrictive speed limit but stop after the first match (sounds extra weird, but this is kind of what the last yaml with the if else looks like, see below)

In Java it would be this:

```java
// 1
if (road_class == MOTORWAY)
   return 100;
if (surface == GRAVEL)
   return 70;
if (road_environment == TUNNEL)
   return 80;

// 2
double result;
if (road_class == MOTORWAY)
   result = 100;
if (surface == GRAVEL)
   result = 70;
if (road_environment == TUNNEL)
   result = 80;
return result;

// 3
double maxSpeed = encoder.maxSpeed();
if (road_class == MOTORWAY)
    maxSpeed = Math.min(maxSpeed, 100);
if (surface == GRAVEL)
    maxSpeed = Math.min(maxSpeed, 100);
if (road_environment == TUNNEL)
    maxSpeed = Math.min(maxSpeed, 100);

// 4
double maxSpeed = encoder.maxSpeed();
if (road_class == MOTORWAY)
    maxSpeed = Math.min(maxSpeed, 100);
else if (surface == GRAVEL)
    maxSpeed = Math.min(maxSpeed, 100);
else if (road_environment == TUNNEL)
    maxSpeed = Math.min(maxSpeed, 100);
```

I guess the problem is that the logic we see in Java here gets 'lost' when converting this to yaml, or at least we have to be careful to make the yaml clear, but let's decide which Java logic we want first... And maybe we even need another option than the four I mentioned, e.g. combinations of these.

Reply:
Btw for speed/priority I think its more clear, because in this case we simply want the product of all matching rules, but let's discuss this after `max_speed` is clear.

Reply:
To me 1) and 3) both make sense. 

Number 1) is very easy to understand. there can be cases where some rules are overruled by others. not even sure if this is a good or bad thing.

In yaml it would be *literally*:

```yaml
max_speed:
 - if: road_class == MOTORWAY
   value: 100
 - else if: surface == GRAVEL
   value: 70
 - else if: road_surface == TUNNEL
   value: 80
```

Number 3) is what we do in master, but gets very confusing when we use the format with `if` in yaml I think.

For 3) a possible yaml could look like this:

```yaml
max_speed:
  - condition: road_class == MOTORWAY
    value: 100
  - condition: surface == GRAVEL
    value: 70
  - condition: road_environment == TUNNEL
    value: 80
```

Avoiding `if` (or at least avoiding if with the various possible combinations with else) hopefully makes it clear that all conditions are equivalent and maybe also separates this from the (different) logic we use for speed/priority. We could even make it extra clear like this:

```yaml
max_speed:
   minimum_of:
     - condition: road_class == MOTORWAY
       value: 100
     - condition: surface == GRAVEL
       value: 70
     - condition: road_environment == TUNNEL
       value: 80
```

Ok and now that we are at it: How to handle max speed fallback in these two cases? For 1) its very simple: we just add

```yaml
max_speed:
   # see above
   ...
   - else: 90
```

For 3) its also easy, just add it as special value (works better if we have the `minimum_of` field)

```yaml
max_speed:
   minimum_of:
       # see above
       ...
   otherwise: 90
```

Reply:
Currently we allow an arbitrary combination of 3+4: if-then-else-if blocks (aka `first_match`) and also just a collection of if-clauses (default). To play with this and make this clear I implemented the new syntax :) . To me this feels natural (1:1 to Java) and we do not have something different for the max_speed compared to the factors. I.e. it is `speed = speed <op> value` where for factors it is `<op>` = `multiply` and for max_speed it is `<op>` = `min`.

Now when it comes to the fallback max_speed I think we only need to loop through all if blocks and pick the minimum value and within one if block (including else-if and else) we pick the maximum. If one if block has no else we can skip it i.e. use the flagEncoder max_speed. A special case is "if: true" and "if: false".

Reply:
So this

```yaml
max_speed:
  - if: road_class == MOTORWAY
    then: 100
  - else if: surface == GRAVEL
    then: 70
  - else if: road_environment == TUNNEL
    then: 80
```
 means:

```java
double maxSpeed = encoder.getMaxSpeed();
if (road_class == MOTORWAY)
   maxSpeed = Math.min(maxSpeed, 100);
else if (surface == GRAVEL)
   maxSpeed = Math.min(maxSpeed, 70);
else if (road_environment == TUNNEL)
   maxSpeed = Math.min(maxSpeed, 80);
```
?

Reply:
Yes

Reply:
> Currently we allow an arbitrary combination of 3+4: if-then-else-if blocks
What do you mean 3+4?

Reply:
> What do you mean 3+4?

You had these examples 3) and 4). Where 3) had three if-blocks and 4) had one if-then-else-if block.

> And where would do you think we should allow 'else'?

Do you mean why we should allow `else`? Why not? So e.g. a truck should get a max speed of 100km/h for primary and motorway roads but only 90km/h for everything else.

Reply:
> You had these examples 3) and 4). Where 3) had three if-blocks and 4) had one if-then-else-if block.

Oh ok :) I thought this means: "currently we allow an arbitrary combination of 7 if-then-else-if blocks... but nevermind ;)

> Do you mean why we should allow else?

No, I was just trying to find out how it is used, e.g. like this:

```yaml
- if: road_class == MOTORWAY
  then: 100
  else: 50
```

or this:

```yaml
- if: road_class == MOTORWAY
  then: 100
- else: 50
```

But I guess you mean the latter? And do you want to allow `else` without preceding if?

Reply:
To me this syntax is still confusing, because I just keep reading 'then' (and it was even worse with 'value') as **return**. 

What would not be confusing to me would be this:

```yaml
max_speed:
   - if: road_class = MOTORWAY
     reduce_to: 100  
   - else if: surface == GRAVEL
     reduce_to: 70
   - else:
     reduce_to: 80
```

```yaml
priority:
  - if: road_class == MOTORWAY
    multiply_with: 0.5
  - else if: surface == GRAVEL
    multiply_with: 0.3
  - else:
    multiply_with: 0.9
```

Reply:
> But I guess you mean the latter? 

Yes, every list entry has only a single operation and so a single number.

> And do you want to allow else without preceding if?

The checking is done from janino :)

Basically every Java block requires a new list entry via `-`.

> And I see it as an advantage not a drawback that we have different keywords for max_speed and priority/speed

Indeed, this reads better. Maybe replace the underscores with spaces similar to `else if`?

Reply:
> Maybe replace the underscores with spaces similar to else if?

Either way is fine for me, but lets make it consistent: Either use space for 'else if' and 'multiply with' or use underscores for both. As you seem to prefer space lets take this (I did not think about it when I put the underscore).

Do you think its a problem that `reduce to` is a no-op for example when 'reducing' 60 to 90? `minimize_to` might be a bit more precise, but maybe also too technical. I'm a bit undecided, but I think `reduce to` sounds good.

> The checking is done from janino :) Basically every Java block requires a new list entry via -.

Does this mean an else block without if is not possible, because in Java we cannot use a single else

```yaml
# invalid
max_speed:
  - else: 65

# valid
max_speed:
  -if: true
   reduce_to: 65
```
   
Ok I think now we are converging, so let's talk about the max speed fallback again :)

> Now when it comes to the fallback max_speed I think we only need to loop through all if blocks and pick the minimum value and within one if block (including else-if and else) we pick the maximum. If one if block has no else we can skip it i.e. use the flagEncoder max_speed. A special case is "if: true" and "if: false".

Yes, I think you are right. For every 'statement' (if/else if/else chain) that has an 'else' (and therefore applies to *all* edges) we take the maximum and out of these maxima the minimum. This way we can be sure that the speed is never higher than the value we obtain and that we chose the max speed fallback as tight as possible.

Sorry for the little detour, I just kept reading value/then as 'return' and got all confused about this. But now the format makes a lot of sense to me.

Reply:
I just thought about this and now we could even merge speed_factor and max_speed! Like:

```yml
speed:
 - if: road_class == TERTIARY
   multiply with: 0.8
 - if: surface != ASPHALT
   reduce to: 100    
```

> Do you think its a problem that reduce to is a no-op for example when 'reducing' 60 to 90

No. 

I'm not yet 100% satisfied with the name. I like `minimize to` or maybe `lower to` ... not sure.

> Does this mean an else block without if is not possible, just like in Java we cannot use a single else?

Yes

> we take the maximum and out of these maxima the minimum
> This way we can be sure that the speed is never higher than the value we obtain and that we chose the max speed fallback as tight as possible.

:+1: 

> Sorry for the little detour,

It was worth a lot I guess :) 

> I just kept reading value/then as 'return' and got all confused about this. But now the format makes a lot of sense to me.

Yes!

Reply:
> I just thought about this and now we could even merge speed_factor and max_speed! Like:

Oh interesting yes. But maybe we have to be a bit careful, because then the order becomes important again? For example `100 multiply with 0.8 reduce to 50 = 50`, but `100 reduce to 50 multiply with 0.8 = 40`. Hm, not sure if this is easier or harder to use this way. Having all speed-related modifications in one section also seems nice.

> I'm not yet 100% satisfied with the name. I like minimize to or maybe lower to ... not sure.

A few other options would be: `max`, `set max`, `limit` or `limit at`, especially if we remove the speed category.

```yaml
speed:
- if: road_class == MOTORWAY
  limit: 70
```

We probably should find out whether or not we keep the `max_speed` section or not first.

I also thought there might be room for other operators, for example setting the speed to a certain value instead of multiplying the default speed with some factor. Something like a `set to` operator. But maybe we should just make sure the format is flexible enough and decide this later.

Maybe it would be more generic to use the operator as value instead of key:

```yaml
priority
- if: road_class == MOTORWAY
  operator: multiply_with
  value: 0.8
```

Its more verbose and a bit harder to read, but I guess it is more generic so the datatype stays the same for the different operators.

Reply:
> Oh interesting yes. But maybe we have to be a bit careful, because then the order becomes important again?

Ah, indeed. Hmmh, we could argue that this is the responsibility of the user and only document this?

> Something like a `set to` operator. 

Yes :+1: 

> A few other options would be: `max`, `set max`, `limit` or `limit at`

Interesting & nice :)

> Its more verbose and a bit harder to read, but I guess it is more generic so the datatype stays the same for the different operators

Yes, it is harder to read. Can you explain the advantage you see - why would this be more generic?

Reply:
> Ah, indeed. Hmmh, we could argue that this is the responsibility of the user and only document this?

Yes its more responsibility for the user, but that is also why I am wondering if combining everything into `speed` is really an improvement. An alternative could be a `speed` section with two sub-sections: `speed_factor` and `max_speed`?

> Yes, it is harder to read. Can you explain the advantage you see - why would this be more generic?

Not so many advantages. The datatype behind this could be more generic, but to me readability seems more important here. And to make the datatype fully generic we would also have to put the `if' key as value of some `clause` key. This was just a thought, as far as I am concerned we can just keep `multiply with: 0.8`.

Reply:
> Yes its more responsibility for the user

Hmmh maybe not only more responsibility but maybe could it cover more use cases? Also with separate sections the user would need to study the docs on how we do this and with his own code it would be obvious. But yeah ... not sure.

> An alternative could be a speed section with two sub-sections: speed_factor and max_speed?

I would keep it as it is or merge it into just one `speed`-section :)

Reply:
Let's merge it into one speed section. As long as the logic is as simple as 'apply the operators with matching conditions from top to bottom' this should be easy enough. For example if users want the previous behavior where the max_speed limitations are applied after the speed factors were calculated they can simply move all clauses with `reduce_to` to the bottom of the `speed` section.

Btw we could now also use other clauses than `reduce_to` to find the max speed fallback, but since this is just an optimization we can probably postpone this.

Reply:
One problem could be that merging the speed sections makes merging custom models harder? Same with adding more operators maybe?

Reply:
> Let's merge it into one speed section

Ok :) 

> Btw we could now also use other clauses than reduce_to to find the max speed fallback, but since this is just an optimization we can probably postpone this.

Yes and yes, let's apply the max speed fallback to `reduce to`.

And shall we rename this to `limit to`?

> One problem could be that merging the speed sections makes merging custom models harder? Same with adding more operators maybe?

Merging custom models is currently very simplistic: we just append the expressions. This is easy to understand and should also never lead to a bigger speed or priority as long as we check the values for every operator (I think?).

The only downside is that it might be a lit bit inefficient e.g. if the server-side model defines 
`if road_class == PRIMARY multiply with 0.5` 
and the query comes with 
`if road_class == PRIMARY multiply with 0.8` 
we could create an optimized model with
`if road_class == PRIMARY multiply with 0.4` 
instead of two separate expressions.

Reply:
> And shall we rename this to limit to?

Yes

> Merging custom models is currently very simplistic: we just append the expressions. This is easy to understand and should also never lead to a bigger speed or priority as long as we check the values for every operator (I think?).

Ok. Yes if alll factors are < 1 the overall factor will be < 1 as well (or what are you worried about?). I am more concerned about a server side model with some `multiply_with` and some subsequent `limit to` rules. If we append a similar client model we will apply the factors to the already limited speeds/priorities? Simply appending the models is easy to understand, but not sure if this is powerful enough? For example should it be possible to 'revert' or override a rule?

Reply:
> If we append a similar client model we will apply the factors to the already limited speeds/priorities? 

Yes. This makes the most sense to me.

> For example should it be possible to 'revert' or override a rule?

For client-side models I would say no (e.g. the LM preparation would not be usable).

And for extending a server-side model on the server-side then yes, the current approach is not sufficient. But currently this is not supported anyway. But I'm not sure if we should make reverting or overriding rules possible at all. Imagine we have a truckA profile with a 90km/h speed limit and we want a new truckB profile that extends truckA but increases this speed to 100km/h. To solve this we can use truckA with 90km/h to extend truckA with 100km/h, which would also make it possible to use the same LM preparation. This won't work if you want one property stricter and another property less strict, but in such a case copying all the rules into a third base profile truckC could work (and then extending truckA from truckC and truckB from truckC). Do you know what I mean :) ?

Reply:
> Yes. This makes the most sense to me.

Ok its probably simple to implement and easy to understand.

> For client-side models I would say no (e.g. the LM preparation would not be usable).

Ok

> And for extending a server-side model on the server-side then yes, the current approach is not sufficient. But currently this is not supported anyway.

Ok I was just wondering if the changed format provides a problem to do this in the future.

> Do you know what I mean :

I'm not sure :rofl: 

> To solve this we can use truckA with 90km/h to extend truckA with 100km/h, which would also make it possible to use the same LM preparation

You said truckA has speed limit 90km/h, but what do you mean 'we can use truckA with 90km/h to extend truckA with 100km/h'? And why can we use the same LM preparation when we increase the speed (*reduce* the weight)? Isn't it the other way around: Re-using the same landmarks is only possible as long as we increase the weights?

To extend custom models on the server side I think LM compatibility is less important, because we can always enforce using certain landmarks (at our own responsibility) using `preparation_profile` and even more so because the custom models shall be used with AStar or CH as well. This is more about how we can share code/yaml between custom models for example when we have a complex configuration and need multiple similar variants of it. But we can probably solve this with some kind of 'include' mechanism just like you proposed: truckA and truckB use some common configuration from truckC (I think I understood this part).

Reply:
> we can use truckA with 90km/h to extend truckA 

Ups, I messed up my example :D ... I meant: we have truckB with 100km/h as base and now we want truckA with the same properties (i.e. extending truckB) but with 90km/h, i.e. increasing the weight.

> To extend custom models on the server side I think LM compatibility is less important, because we can always enforce using certain landmarks (at our own responsibility) using preparation_profile and even more so because the custom models shall be used with AStar or CH as well. 

Yes, server-side models do not need to have these limitations also because we could create a separate LM preparation or allow a different preparation_profile at our own responsibility.

> This is more about how we can share code/yaml between custom models for example when we have a complex configuration

Yes. Looking into other routing engines this does not seem to be a solved thing and they copy&paste the config.

I thought a bit about this how we could do this and one way to approach this is to explicitly name a set of if-statements ("define a method") so that it can be included later (in the same model and) externally:

```yaml
export:
- avoidSmallRoads:
  - if: road_class == LIVING_STREET || road_class == TERTIARY || road_class == RESIDENTIAL
    multiply with: 0.5

priority:
- run: avoidSmallRoads()
```

This way it is up to the one who writes the new custom model if it is included before or after some other statements. Also in theory a method could be used for `speed` and `priority` although it was intended only for `priority`. Not sure if we need to limit this.

Reply:
Should we already introduce this nesting to make the configs forward compatible? I mean just push the if statements one level down like you did with `avoidSmallRoads` to introduce a 'namespace'? Then again we can also later introduce the 'default' namespace that uses no nesting. Having such namespaces we could also specify what to include/exclude when merging or re-using models. This seems to be a rather complex topic, but just giving names to single or a group of `Statement`s probably makes this a lot easier, regardless about how we will do this in detail.

Reply:
I would prefer it if only certain statements are usable in "subclasses" not automatically all statements of `priority` and `speed` (is this what you meant with 'default' namespace?). This way the super class is only responsible for those "exported" statements and also does not need to give everything a name.

Reply:
Ok I see its more comfortable if we do not always have to specify a name. That is what I meant with 'default' namespace: It is allowed to put statements without nesting and choosing a label. But if we ever wanted to 'import' or exclude such statements in the future we could define that importing 'default' means we import everything from the top level.

Why do you think it would be better if super classes had to explicitly 'export' something? 

Another thing we should consider is that currently `speed` and `priority` are lists in yaml, but what if add the labels or need to add some special keys like: 

```yaml
speed:
   # do we need this someday?
   parent: ../some/other/car.yml
   - if: ... # as before
   # here comes a labeled group of if statements
   avoidSmallRoads:
       - if: #...
 ```
Do you think we can keep `speed/list` as list, but simply parse the single entries as different types, i.e.:

```yaml
speed:
   - parent: #...
   - if: #...
   - avoidSmallRoads: #...
```

Reply:
> Why do you think it would be better if super classes had to explicitly 'export' something?

If something is not intended for extension IMO it should not be extended or reused. Otherwise the profile author cannot make changes in `speed` or `priority` as it will likely brake something in extending profiles. (for the same reason I would prefer that the default of a class should be `final` ... unfortunately it isn't in Java :))

> do we need this someday?
> parent: ../some/other/car.yml

A better notation would be probably to include a more explicit pointer to the statement list directly like `parent: car->speed` or something? Because otherwise which statements should it import, from `speed` or `priority`? And due to the reason above I would allow pointing only to named statements explicitly intended for reusage like `car->avoid_bridges`.

> Ah ok, but with the export/run syntax maybe this is already solved.

Yes

> Maybe just rename run -> include then including just means text replacement.

Yes, sounds good :+1:

> Are you planning to add this to this PR already?

IMO this PR is already quite big and I would prefer to fix the missing bits. And maybe introduce the extension possibility in the next version. (One important feature before should be probably to define custom models without a separate file.)

Reply:
Sounds good!

> (for the same reason I would prefer that the default of a class should be final ... unfortunately it isn't in Java :))

Java 9 modules would allow being more explicit what does and what does not belong to the 'public' API.